### ISSUE TITLE
Support the MLX engine (safetensors models) on the K80 fork

### ISSUE THREAD

**Issue by dogkeeper886** [dogkeeper886/ollama37#187]
## Goal
Run MLX-format models on this K80 fork — not just GGUF. Today an `-mlx` tag routes to an engine our fork doesn't ship. Porting it lets us run MLX-only releases (e.g. the Qwen3.6 coding fine-tunes that publish no GGUF) on K80.

This is a tracking/epic issue. A porting plan (sub-issues) follows.

## Background — upstream behavior (traced from code)
Full trace: `docs/traces/mlx-vs-gguf-routing.md`.

- **Routing is one manifest field.** `IsMLX()` returns `Config.ModelFormat == "safetensors"` (`server/images.go:80`). GGUF → llama.cpp/GGML engine; safetensors → MLX engine spawned as a subprocess `ollama runner --mlx-engine` (`server/sched.go:436/458`, `x/mlxrunner/client.go:292`). Same architecture (`qwen3_5`) underneath — only storage format + runtime engine differ.
- **MLX is NOT Apple-only.** `CheckPlatformSupport()` allows `linux`/`windows` (`x/imagegen/memory.go:34`), and `client.go:296-348` wires `LD_LIBRARY_PATH`, globs `mlx_cuda_*`, and points MLX's JIT compiler at CUDA headers (`CUDA_PATH`/`CUDA_HOME`). MLX has a real **CUDA backend**.
- **Our fork ships none of it.** `x/mlxrunner` + `x/imagegen/mlx` are absent here (grep: 0 files).

## Porting scope (to break into sub-issues)
1. **Port `x/mlxrunner`** — runner/server/client + the `mlx` cgo bindings — into the fork.
2. **Wire routing** — bring the `IsMLX()` safetensors check + `mlxrunner.NewClient` branch into our (possibly diverged) `server/sched.go`.
3. **MLX CUDA backend libs** — build/bundle `mlx_cuda_*` and integrate into our build + CI (`test-build.yml`).
4. **Compute capability 3.7 (sm_37)** — make MLX-CUDA emit Kepler kernels. This is *work*, the same kind we already did for llama.cpp on this fork — trace MLX-CUDA's arch flags / JIT target and patch as needed. Not treated as a blocker.
5. **Smoke test** — load a safetensors model end-to-end on K80, coherent output, clean unload (mirror the TC-MODELS pattern).

## Open questions (to resolve during porting)
- MLX-CUDA's arch-flag / JIT mechanism and where to inject sm_37.
- MLX cgo build deps and CI footprint on the K80 runner.
- How much our `sched.go` / runner plumbing has diverged from upstream.

## References
- `docs/traces/mlx-vs-gguf-routing.md` — upstream routing + platform trace
- Related to #170 (Qwen3.6 feasibility)
- Related to #180 (MLX→GGUF conversion — the alternative path, closed; this issue is the *native* path)

## Dependencies
- None (epic). Sub-issues to follow from the porting plan.

**Comment by dogkeeper886** [comment]
## Porting plan (sub-issues)

Broken into 5 ordered, dependency-linked sub-issues:

- [ ] #188 — Build MLX library + CUDA backend targeting compute 3.7 (CUDA 11.4) — **foundation**
- [ ] #189 — Port `x/mlxrunner` (+ required `x/models` MLX impls) — depends #188
- [ ] #190 — Wire safetensors → mlxrunner routing into the scheduler — depends #189
- [ ] #191 — Integrate MLX libs into the build + CI — depends #188
- [ ] #192 — End-to-end smoke test: safetensors model on K80 — depends #190, #191

```
#188 (MLX+CUDA sm_37 build)
 ├─► #189 (port mlxrunner) ─► #190 (routing) ─┐
 └─► #191 (build/CI) ─────────────────────────┴─► #192 (smoke test)
```

Start point: #188 (gates everything else).

**Comment by dogkeeper886** [comment]
## Status + revised plan (post-Phase-B link milestone)

#188 is closed. `libmlx.a` archives on sm_37 / nvcc 11.4 / gcc-10, `qwen_smoke` runs an MLX forward-op surface (matmul + RMS-norm + RoPE + softmax + SDPA + conv1d) on the K80 end-to-end via the standardized `test-mlx-smoke.yml` workflow.

### Target sharpened: `qwen3.6:35b-mlx` (A3B / MoE + DeltaNet)

The epic was deliberately vague ("run MLX-format models"). After the Phase-B milestone and a target review, the concrete first model is:

**Ollama library: `qwen3.6:35b-mlx`** — 22 GB, MLX format, architecture `Qwen3_5MoeForConditionalGeneration` (DeltaNet hybrid + MoE expert FFNs).

Why this and not e.g. `qwen3.6:27b-mlx`:
- A3B (active-3B-per-token MoE) is the architecture class where engine choice genuinely matters — combined DeltaNet recurrence × MoE routing is exactly where naive implementations leave performance on the table, so MLX's graph compilation has a fair chance of doing something interesting. Dense models the GGML path already runs well (qwen3.5/qwen3.6) don't have that opening.
- It's the riskiest target (DeltaNet × MoE × multi-GPU sharding) — so it's the one whose validation result actually decides whether the MLX path earns its keep.

### Two discoveries that change the plan

1. **No GGUF→MLX converter needed.** Ollama publishes MLX-format weights directly via their library — `ollama pull qwen3.6:35b-mlx` lands a safetensors-shaped 22 GB blob ready to load. Deletes the converter-tooling workstream that #192's "pull or convert" AC contemplated.
2. **Runtime image already has CUDA toolkit libs bundled.** `/usr/lib/ollama/libcudart.so.11`, `libcublas.so.11`, `libcublasLt.so.11`. No Dockerfile change needed for #191 — the smoke run linked cleanly with zero `ldd_missing`.

### Likely real blocker: VRAM topology

22 GB > 12 GB per K80 die. The K80 host has 3 K80 cards (24 GB pooled), so multi-die layer sharding (P2P over PCIe Gen3 x16) is the realistic deployment. Or layer offload. Either way, this is the part with no upstream pattern to copy (MLX is overwhelmingly designed assuming everything fits on one device) and probably dominates the remaining engineering effort, not the kernel ports.

### Revised sub-issue scope

| Issue | Original scope | A3B-retarget scope |
|---|---|---|
| #189 — port `x/mlxrunner` + `x/models` MLX impls | generic | A3B-specific: `x/models/qwen3_6_a3b` (DeltaNet + MoE), Go cgo wrapper around libmlx.a, tokenizer (HF), sampling, chat templates |
| #190 — safetensors → mlxrunner routing | unchanged | unchanged |
| #191 — build/CI integration | unchanged | **partially done** — `test-mlx-smoke.yml` covers AC #3 (CI builds MLX-enabled). Remaining: runtime-image bundling of `libmlx.a` (AC #1, #2, #4). Bonus: AC #1's `mlx_cuda_*` directory layout no longer needed (we link statically into the runner exe, not via dlopen). |
| #192 — end-to-end smoke | "qwen3_5 MLX build" | retarget to `qwen3.6:35b-mlx`; pull during download window (22 GB) |

### Kernel punch-list, A3B-specific

Currently throw-stubs in `k80_runtime_stubs.cpp` that A3B will trip:

| Stub | Need real impl? | Reason |
|---|---|---|
| `affine_dequantize` | **yes** | every quantized linear in A3B (DeltaNet projection + MoE expert FFNs) |
| `Scan::eval_gpu` | **yes** | DeltaNet recurrence chunking |
| `Gather`/`Scatter`/`SliceUpdate` | **yes** | KV cache (attention layers) + DeltaNet state |
| `cutlass_gather_mm` reroute | **yes** | MoE expert routing |
| `cutlass_grouped_gemm_unaligned` reroute | **yes** | MoE expert FFN dispatch |
| `CublasGemm::run_batched` | **yes** | batched GEMM in attention path |
| `qmv`/`qmm_*` family | **no** — supports_*=false makes dispatch skip them, falls back to dequant+matmul | (already wired this way) |
| `FFT`/`Hadamard`/`MaskedScatter`/`QQMatmul`/`ConvertFP8` | **no** | not on A3B inference path |
| `distributed::*` | **no** | single-host |
| `apply_block_mask`/`copy_with_block_mask`, JIT runtime | **no** unless something surprises | off-path |

### Phase plan from here

- **Phase C** (issues #189, #190): runner exe + scheduler routing.
  - C.1: port `affine_dequantize` (smallest first real kernel; unblocks every quantized linear).
  - C.2: `cutlass_gather_mm` + `cutlass_grouped_gemm_unaligned` → cuBLAS-per-expert.
  - C.3: `Gather`/`Scatter`/`SliceUpdate` real impls.
  - C.4: `Scan` real impl.
  - C.5: `x/mlxrunner` Go cgo wrapper + `x/models/qwen3_6_a3b`.
  - C.6: `IsMLX()` routing in scheduler (#190).
- **Phase D** (issue #192): end-to-end run.
  - D.1: VRAM sharding / multi-K80 layer placement.
  - D.2: pull `qwen3.6:35b-mlx`, load safetensors, run forward.
  - D.3: `/api/chat` returns coherent output.
  - D.4: TC-MODELS-style YAML test case.

### Realistic re-estimate

Phase C+D for A3B target: probably **4–8 weeks** of focused work, dominated by Phase D's sharding (no upstream pattern) and Phase C.5's runner port. The kernel reroutes (C.1-C.4) are well-understood; the unknowns are in the runner + memory topology.

If at any Phase-D milestone the result is "MLX-on-K80 isn't better than GGML-on-K80 on any metric we care about," the disciplined call is to land the work as a `frozen/` reference and not maintain it — the merged commits remain a credible evaluation outcome.

**Comment by dogkeeper886** [comment]
## A3B model inspection findings — revised Phase C plan

While the download is running, I inspected `mlx-community/Qwen3.6-35B-A3B-4bit` (the HF mirror — see note on Ollama gate below) to put the kernel punch-list on a data-driven basis. Trace: `docs/traces/qwen36-a3b-mlx-inspection.md` (commit `11d42bba`).

### Discoveries that change the plan

1. **Ollama library gates MLX models on OS.** `ollama pull qwen3.6:35b-mlx` returns `412: this model requires macOS`. Our fork's whole point is to break that assumption, but the gate is server-side, not client. Workaround: pull from `mlx-community` HF mirror (19 GB, 4 shards, standard MLX safetensors layout). Runner (#189) points at the local dir directly — no need to round-trip through ollama's manifest system.

2. **Model is multimodal**, with a 27-block vision_tower (327 of 2090 tensors). Text-only inference ignores it.

3. **DeltaNet hybrid + MoE.** 40 language-model layers split into 30 `linear_attn` (DeltaNet recurrence) + 10 `self_attn` (standard attention). Memory note `[[project-mlx-on-k80]]` already flagged DeltaNet + MoE as the combined architecture — now confirmed at the tensor level.

4. **MoE topology**: every layer has `mlp.gate` (router, 8-bit) + `mlp.shared_expert` (always-on FFN) + `mlp.switch_mlp` (sparse routed FFN, single tensor stacked over experts) + `mlp.shared_expert_gate` (8-bit). The shared expert is a permanent residual; sparse routing only varies the switch_mlp call.

5. **Quantization is mixed precision** — default `affine/bits=4/group=64`, with per-tensor `bits=8` overrides on every MoE router gate. All within the `{2,4,8}` regular cases that my `affine_dequantize` (#196) and dequant+cuBLAS fallback (#197) already handle. The runner has to read `config.json` per-tensor before constructing each `QuantizedMatmul`; can't assume uniform bits.

6. **Embed + LM head are quantized too.** So even embedding lookup is `take` → quantized read → dequant. Not just a raw memcpy.

### Punch-list, cross-referenced with k80_runtime_stubs.cpp

| What the model uses | Stub status | Notes |
|---|---|---|
| Quantized linears (Q/K/V/O proj, FFN proj, lm_head) | ✅ **landed** (#196 + #197) | dequant→cuBLAS, exercised in qwen_smoke |
| Standard attention (10 layers: matmul, softmax, rms_norm, rope, sdpa) | ✅ already in libmlx.a | Phase A + Phase B work |
| Embedding lookup (`take` on quantized embed) | ❌ `Gather::eval_gpu` stub | First call any forward pass makes — will trip immediately |
| **DeltaNet recurrence (30 layers)** | ❌ `Scan::eval_gpu` stub | State-space chunks; upstream uses NVRTC/JIT-built kernel |
| **MoE expert routing** | ❌ `cutlass_grouped_gemm_unaligned` + `cutlass_gather_mm` stubs | Reroute to cuBLAS-per-expert; needs Gather/Scatter in place too |
| KV cache reads/writes | ❌ `Gather`/`Scatter`/`SliceUpdate` stubs | Standard attention path |

### Why I'm pausing speculative kernel ports

Upstream's `Gather::eval_gpu` is deeply tied to the **NVRTC/JIT subsystem** (`mlx/backend/cuda/jit_module.h` + generated `cuda_jit_sources.h`) — it generates kernel names dynamically per `dtype × NIDX × IDX_NDIM × LocT` combination. To port without JIT, every used combination needs a static template instantiation. Same shape for `Scan`, `cutlass_*`, `Scatter`. Without the runner driving the forward pass, we don't know which instantiations matter — speculative porting wastes effort.

**Discipline**: let the runner trip stubs in priority order at runtime, port the actual hit cases with concrete shapes known.

### Revised Phase C priority

| Old | New |
|---|---|
| C.1: `affine_dequantize` port | ✅ done (#196) |
| C.2: `cutlass_gather_mm` + `cutlass_grouped_gemm_unaligned` → cuBLAS-per-expert | **deferred** — port when runner trips them |
| C.3: `Gather`/`Scatter`/`SliceUpdate` real impls | **deferred** — port when runner trips them |
| C.4: `Scan` real impl (DeltaNet recurrence) | **deferred** — port when runner trips it |
| C.5: `x/mlxrunner` Go cgo wrapper + `x/models/qwen3_6_a3b` | **now critical path** — drives the rest |
| C.6: `IsMLX()` routing in scheduler (#190) | unchanged |

The runner is now the dominant remaining work, and the kernel ports become reactive (driven by runtime trip data). Resets the realistic ordering for #189 to start before further #188-style kernel PRs.

### Memory topology — the actual hard part

A3B 4-bit is 19 GB; one K80 die is 12 GB. But **active params per token are ~3 B** — sparse MoE means most expert tensors are cold on any single forward pass. Possibilities:
- Multi-die layer placement across 3 K80s (24 GB pooled, P2P over PCIe Gen3 x16). Layer-wise sharding by reference frequency.
- CPU↔GPU expert paging: hot weights (embed, attn, shared experts, router, last-N hot routed experts) stay GPU-resident; cold routed experts stream from host RAM per token.
- The shared_expert always runs — that's at least one MoE FFN's worth of weights pinned.

This is the part with no upstream pattern. Probably dominates Phase D effort.

**Comment by dogkeeper886** [comment]
Parking the MLX-on-K80 initiative — decision to **stop** MLX format support.

**Rationale.** The goal was to reach capable models on the K80 that the GGML engine couldn't. That gap has effectively closed:

- **MoE runs well on GGML** — `gpt-oss:20b` sustains ~11–15 tok/s on the K80 (GGUF/GGML, no MLX involved).
- **DeltaNet runs on GPU on the K80** — the CUDA kernels (#18, #19: softplus/cumsum/tri/solve_tri/fill/diag) are merged and registered in the ggml-cuda dispatch, and the qwen3.5 DeltaNet graph builder (#22) produces coherent GPU output. There is no CPU fallback in that path.

MLX would add **format reach only** (loading safetensors-only models), **not speed** — on the K80 the hardware is the performance ceiling, and an MLX-CUDA dequant→cuBLAS path will not beat tuned GGML kernels. So the cost/benefit no longer justifies standing up a second runtime (mlxrunner + the deferred NVRTC/JIT kernel ports + the 19 GB > 12 GB multi-die sharding problem at Phase D).

**What happens to the code.** The work done so far (libmlx.a build + `qwen_smoke` link milestone, #188) stays **frozen in-tree** under `x/mlx/` and `x/mlxrunner/`. It is fully isolated and costs nothing to keep: not built into the runtime image, not referenced in the inference path, and its only CI (`test-mlx-smoke.yml`) is `workflow_dispatch`-only.

**Reopen if** a target model is distributed MLX/safetensors-only **and** can't be converted to GGUF — that's the one scenario where MLX would still be the only path.

Closing #187, #189, #190, #191, #192 as not planned. (#188 is already complete.)
